### PR TITLE
Halstead complexity measuring

### PR DESCRIPTION
**Pull request by bragov4ik**
Halstead operators/operands counting
Formulae for complexity calculations
Integration of the measurements into kys package
+minor fixes

### CODE REVIEW

File: pkg/halstead/halstead.go [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+	operators map[string]uint
+	operands  map[string]uint
+}
+
+func NewHalsteadInfo() *HalsteadInfo {
+	return &HalsteadInfo{
+		operators: make(map[string]uint),
+		operands: make(map[string]uint),
+	}
+}
```
Comment (i1i1) [inline pkg/halstead/halstead.go:20]: Do we need constructor though?

File: pkg/halstead/halstead.go [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+	n1Tot, n2Tot := float64(info.getN1Total()), float64(info.getN2Total())
+	D := (n1Dist * n2Tot) / (2 * n2Dist)
+	V := (n1Tot + n2Tot) * math.Log2(n1Dist+n2Dist)
+	return D * V
+}
+
+func (info *HalsteadInfo) addToken(nextToken token.Token) {
+	tokenName := fmt.Sprintf("token:%s", nextToken.String())
+
+	NOT_OPERATORS := []token.Token{token.ILLEGAL, token.EOF, token.COMMENT, token.IDENT, token.INT, token.FLOAT, token.IMAG, token.CHAR, token.STRING}
```
Comment (i1i1) [inline pkg/halstead/halstead.go:93]:
```suggestion
	NOT_OPERATORS := []token.Token{
	    token.ILLEGAL,
	    token.EOF,
	    token.COMMENT,
	    token.IDENT,
	    token.INT,
	    token.FLOAT,
	    token.IMAG,
	    token.CHAR,
	    token.STRING,
	}
```

File: pkg/halstead/halstead.go [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+	info.addToken(token.RETURN)
+}
+
+func (info *HalsteadInfo) AddSelectStmt(node *ast.SelectStmt) {
+	info.addToken(token.SELECT)
+}
+
+func (info *HalsteadInfo) AddSelectorExpr(node *ast.SelectorExpr) {
+	info.addToken(token.PERIOD)
+	return
```
Comment (i1i1) [inline pkg/halstead/halstead.go:319]: No need to do explicit returns at the end of function

File: pkg/halstead/halstead.go [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+	tokenName := fmt.Sprintf("token:%s", nextToken.String())
+
+	NOT_OPERATORS := []token.Token{token.ILLEGAL, token.EOF, token.COMMENT, token.IDENT, token.INT, token.FLOAT, token.IMAG, token.CHAR, token.STRING}
+	
+	if !tokenInArr(nextToken, NOT_OPERATORS[:]) {
+		info.operators[tokenName] += 1
+	}
+}
+
+// TODO count commas
```
[the remark this answers is not in the document]
Comment (bragov4ik) [reply pkg/halstead/halstead.go:110]: +, they're not trivial to count
I thought consider any list inside a node. However, it also covers other cases such as statements inside any block (`{ }`), which do not require any commas.

File: pkg/kys/kys.go [hunk 5]
```diff
@@ -15,6 +17,11 @@ type Info struct {
 	CallExpr   uint
 	AssignStmt uint
 	Comments   uint
+	Halstead   halstead.HalsteadInfo
+}
+
+func NewInfo() *Info {
+	return &Info{Halstead: *halstead.NewHalsteadInfo()}
```
[the remark this answers is not in the document]
Comment (bragov4ik) [reply pkg/kys/kys.go:24]: Can't call NewHalsteadInfo (which is required, see https://github.com/bragov4ik/go-kys/pull/13#discussion_r724163567) without constructor

File: pkg/halstead/halstead.go [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"math"
+)
+
+type HalsteadInfo struct {
+	operators map[string]uint
+	operands  map[string]uint
```
[the remark this answers is not in the document]
Comment (bragov4ik) [reply pkg/halstead/halstead.go:12]: Думал над этим, не придумал зачем это может понадобиться. Могу сделать публичными, если есть причина

File: pkg/halstead/halstead.go (hunk 3 above)
Comment (bragov4ik) [reply pkg/halstead/halstead.go]: solved in https://github.com/bragov4ik/go-kys/pull/13/commits/eb1dd136033e1864abf86d39912ec873f8277698

File: pkg/halstead/halstead.go [hunk 7]
```diff
@@ -0,0 +1,363 @@
+package halstead
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"math"
+)
+
+type HalsteadInfo struct {
```
[the remark this answers is not in the document]
Comment (i1i1) [reply pkg/halstead/halstead.go:10]: Nope. `hasltead.Info` sounds contextual and fine, while `halstead.HalsteadInfo` has repetitions in name

File: pkg/kys/kys.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (i1i1) [reply pkg/kys/kys.go]: Didnt found a comment though. So what is the reason?

File: pkg/halstead/halstead.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (bragov4ik) [reply pkg/halstead/halstead.go]: Agree, renamed it https://github.com/bragov4ik/go-kys/pull/13/commits/74c0f87c835863cc0566a5752cdf3705c92d8178

File: pkg/kys/kys.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bragov4ik) [reply pkg/kys/kys.go]: Won't it copy all fields when returning from the funciton?

**Pull request merged**